File: server/eda/graph/plotly_theme.py
"""
Shared modern Plotly theme for all EDA graph modules.

Usage:
    from .plotly_theme import apply_modern_theme, MODERN_COLORS

    fig = px.bar(...)
    apply_modern_theme(fig, title="My Title")
"""

# ── Soft, vibrant color palette (Chart.js-inspired) ──
MODERN_COLORS = [
    "#ff6384",  # Soft Pink
    "#36a2eb",  # Sky Blue
    "#ffce56",  # Warm Yellow
    "#4bc0c0",  # Teal
    "#9966ff",  # Soft Purple
    "#ff9f40",  # Orange
    "#c9cbcf",  # Light Grey
    "#7dd3fc",  # Light Blue
    "#f0abfc",  # Lavender Pink
    "#6ee7b7",  # Mint Green
]

FONT_STACK = "Inter, Segoe UI, Roboto, Helvetica Neue, Arial, sans-serif"


def apply_modern_theme(fig, title=None):
    """
    Apply a modern, clean theme to any Plotly figure.

    This updates the figure's layout in-place and returns the figure
    for optional chaining.

    Parameters
    ----------
    fig : plotly.graph_objects.Figure
        The Plotly figure to style.
    title : str, optional
        Override the figure's title text.

    Returns
    -------
    fig : plotly.graph_objects.Figure
    """
    fig.update_layout(
        # ── Typography ──
        font=dict(
            family=FONT_STACK,
            size=13,
            color="#4b5563",
        ),
        # ── Backgrounds ──
        plot_bgcolor="#ffffff",
        paper_bgcolor="#ffffff",
        # ── Color palette ──
        colorway=MODERN_COLORS,
        # ── Axes ──
        xaxis=dict(
            gridcolor="rgba(0,0,0,0.06)",
            gridwidth=1,
            linecolor="rgba(0,0,0,0.08)",
            linewidth=1,
            zeroline=False,
            showline=True,
            automargin=True,
            tickfont=dict(size=12, color="#1f2937", family=FONT_STACK),
            title_font=dict(size=14, color="#111827", family=FONT_STACK),
        ),
        yaxis=dict(
            gridcolor="rgba(0,0,0,0.06)",
            gridwidth=1,
            linecolor="rgba(0,0,0,0.08)",
            linewidth=1,
            zeroline=False,
            showline=True,
            automargin=True,
            tickfont=dict(size=12, color="#1f2937", family=FONT_STACK),
            title_font=dict(size=14, color="#111827", family=FONT_STACK),
        ),
        # ── Hover ──
        hoverlabel=dict(
            bgcolor="rgba(50,50,50,0.85)",
            font_size=13,
            font_family=FONT_STACK,
            font_color="#ffffff",
            bordercolor="rgba(0,0,0,0)",
        ),
        hovermode="closest",
        # ── Legend ──
        legend=dict(
            font=dict(size=13, color="#111827", family=FONT_STACK),
        ),
    )

    # ── Title styling ──
    if title:
        fig.update_layout(
            title=dict(
                text=title,
                font=dict(
                    family=FONT_STACK,
                    size=18,
                    color="#374151",
                ),
                x=0.5,
                xanchor="center",
            )
        )

    return fig

File: client/src/shared/plotlyTheme.js
// Shared Plotly visual theme constants — aligned with Matflow design system (style.jsonc).

export const FONT_STACK =
  "'Inter', system-ui, -apple-system, 'Segoe UI', Roboto, sans-serif";

export const MODERN_COLORWAY = [
  "#0D9488", // Primary teal
  "#5EEAD4", // Primary light
  "#10B981", // Accent green
  "#36a2eb", // Sky Blue
  "#ff6384", // Soft Pink
  "#ffce56", // Warm Yellow
  "#9966ff", // Soft Purple
  "#ff9f40", // Orange
  "#6ee7b7", // Mint Green
  "#f0abfc", // Lavender Pink
];

export const THEME_OVERRIDES = {
  font: {
    family: FONT_STACK,
    size: 13,
    color: "#4b5563",
  },
  plot_bgcolor: "#ffffff",
  paper_bgcolor: "#ffffff",
  colorway: MODERN_COLORWAY,
  hoverlabel: {
    bgcolor: "rgba(50,50,50,0.85)",
    font: { color: "#ffffff", size: 13, family: FONT_STACK },
    bordercolor: "rgba(0,0,0,0)",
  },
  hovermode: "closest",
};

export const AXIS_THEME = {
  gridcolor: "rgba(0,0,0,0.06)",
  gridwidth: 1,
  linecolor: "rgba(0,0,0,0.08)",
  linewidth: 1,
  tickfont: { size: 12, color: "#1f2937", family: FONT_STACK },
  title: { font: { size: 14, color: "#111827", family: FONT_STACK } },
  zeroline: false,
  showline: true,
};

export const LEGEND_THEME = {
  font: { size: 13, color: "#111827", family: FONT_STACK },
};

/**
 * Build a themed layout by merging THEME_OVERRIDES + AXIS_THEME + LEGEND_THEME
 * on top of a figure's existing layout. Does NOT touch width/height/margin.
 */
export function applyPlotlyTheme(figureLayout = {}, titleOverride) {
  const backendTitle =
    typeof figureLayout.title === "string"
      ? figureLayout.title
      : figureLayout.title?.text ?? "";

  const titleText = titleOverride || backendTitle;

  return {
    ...figureLayout,
    ...THEME_OVERRIDES,
    font: { ...THEME_OVERRIDES.font, ...figureLayout.font },
    xaxis: { ...AXIS_THEME, ...figureLayout.xaxis },
    yaxis: { ...AXIS_THEME, ...figureLayout.yaxis },
    hoverlabel: { ...THEME_OVERRIDES.hoverlabel, ...figureLayout.hoverlabel },
    legend: { ...LEGEND_THEME, ...figureLayout.legend },
    title: titleText
      ? {
          text: titleText,
          font: { family: FONT_STACK, size: 18, color: "#374151" },
          x: 0.5,
          xanchor: "center",
        }
      : figureLayout.title,
    showlegend: figureLayout.showlegend !== false,
    autosize: true,
  };
}
